### PR TITLE
Expert choice checkpointing

### PR DESCRIPTION
**Pull request by maciek-pioro**
# Description
Changes the code so that it's possible to do checkpointing shenanigans (e.g. reusing a mask in checkpointing's second forward pass). Implements reusing a mask in expert choice.

# Checklist
   - [ ] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
     - Could be divided into two PRs, however I think it would be unnecessary and harmful (it's easier to show the whole concept works in a single PR).
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
[CLICK](https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9aae9724-c7a5-44b3-a894-d7bc9d7bd4df&detailsTab=metadata&shortId=LLMRANDOM-3842&dash=charts&type=run&path=loss_interval&attribute=1&compare=MwNgHCA0CMMKxyA&chartFilter=loss)
      - 3927 is new code + no checkpointing + 2 gpus
      - 3904 is new code + checkpointing + 2 gpus
      - 3853 is new code + no checkpointing
      - 3842 is new code + checkpointing
      - 3687 is old code + no checkpointing
      - 3686 is old code + checkpointing
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] ~If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters~
   - [ ] ~If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)~
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: lizrd/train/train_utils.py [hunk 1]
```diff
@@ -88,12 +86,10 @@ def get_model(
 
     if activation_checkpointing_modules is not None:
         check_fn = lambda x: isinstance(x, activation_checkpointing_modules)
-        non_reentrant_wrapper = partial(
-            checkpoint_wrapper,
-            checkpoint_impl=CheckpointImpl.NO_REENTRANT,
```
Comment (Simontwice) [inline lizrd/train/train_utils.py:93]: what happened to NO_REENTRANT

Comment (Simontwice) [reply lizrd/train/train_utils.py:93]: ah, I see that it's got `use_reentrant = False`. What is the difference?

**Review by kuba-krj — COMMENTED** [review]
So far LGTM; we also discussed the code changes together. Let's just wait for the experiments to finish to make sure everything works and we can merge. Thank you for the work!

File: lizrd/train/train_utils.py (hunk 1 above)
Comment (maciek-pioro) [reply lizrd/train/train_utils.py:93]: That's the same thing.

**Review by kuba-krj — APPROVED** [review]
Just please make sure it works on 2-gpu and we can merge

**Comment by maciek-pioro** [comment]
> Just please make sure it works on 2-gpu and we can merge

@kuba-krj added in the description. Can you take a look and say if you think it looks alright?

**Pull request merged**